stan/math/opencl/copy.hpp

```diff
@@ -100,12 +101,12 @@ inline Eigen::Matrix<double, Eigen::Dynamic, Eigen::Dynamic> from_matrix_cl(
  * Packs the flat triagnular matrix on the OpenCL device and
  * copies it to the std::vector.
  *
- * @tparam triangular_view the triangularity of the source matrix
  * @param src the flat triangular source matrix on the OpenCL device
  * @return the packed std::vector
+ * @throw <code>std::invalid_argument</code> if the matrix is not triangular
  */
-template <TriangularViewCL triangular_view>
 inline std::vector<double> packed_copy(const matrix_cl& src) {
+  check_triangular("packed_copy", "src", src);
```

Comment:
This function seems to only work with triangular matrices. I don't understand what exactley are you proposing here.

Reply:
Sorry, what I'm proposing here is that we make it work for `Entire` instead of throwing. So just an if where if it's `Entire` then we take the vector as is without doing any real unpacking etc.

Reply:
That is just adding functionality we do not need. Actually I belefe throwing is preferable. I can't imagine a use case where someone would need to use `packed_copy()` instead of `from_matrix_cl()` for `Entire` matrix. I guess in vast majority if not all the cases that `Entire` matrix would be passed in `packed_copy()` that would be a bug. Even if we put all data from matrix in the `vector` the user would just have problems later, since he is expecting to get packed triangular matrix as a result.

Reply:
> That is just adding functionality we do not need. Actually I believe throwing is preferable. I can't imagine a use case where someone would need to use packed_copy() instead of from_matrix_cl() for Entire matrix. 

I wouldn't think about it as "is there a use case we can think of" but "should we let users do this". I think if a user wants to do a vector -> matrix_cl or matrix_cl -> vector with `Entire` I don't see a reason we can't allow that.

> I guess in vast majority if not all the cases that Entire matrix would be passed in packed_copy() that would be a bug. 

If we had a check for entire then I'm not seeing how it would be a bug

> Even if we put all data from matrix in the vector the user would just have problems later, since he is expecting to get packed triangular matrix as a result.

One of the major points of this PR is that users don't have to worry about the triangularity. Based on their operations we deduce that for them. They just have to tell us when moving over to OpenCL if the matrix is triangular

Reply:
> a user wants to do a vector -> matrix_cl or matrix_cl -> vector with Entire

Can you think of a use case where this is better or simpler of more convenient to the user than matrix_cl -> MatrixXd? 

Even if there is such a use case we should probably make a separate function that copies from `matrix_cl` to vector (and loses data on the shape of the matrix). I think of `packed_copy()` as a low level optimization that is only used in cases where one knows he has a triangular `matrix_cl` to get only nonzero elements back from GPU faster. In all other cases `from_matrix_cl()` should be used.

And most likely further processing of the data from `packed_copy()` is based on the assumption that this is packed form of triangular matrix. Which makes no sense if it is actually an entire matrix. Such as in `cholesky_opencl.chain()`, which is afaik the only place this function is used.

If we want to have just one function that does everything, than we can implement packed copying functionality within `˙from_matrix_cl()` for the case of triangular matrix. 

> One of the major points of this PR is that users don't have to worry about the triangularity. 

That holds true only as long as data is in `matrix_cl`. We are not changing anything outside it. So triangularity must be specified where user is copying data to GPU and if user does anything triangularity-specific after copying the data back from GPU he must know which triangularity he gets. If his assumption is wrong it is better if he gets a nice error.